Added an article about deployments (#6716)
mesosphere/marathon: docs/articles/deployments.md

# Deployments

> # What is a deployment

Author: that's taken from the template, so I guess we should keep it consistent?

> Every time a user creates a new [service](services.md) modifies or removes an existing one a deployment is created. Basically, deployment is a series of steps that Marathon has to execute in order to fulfill the user request. Let's consider an example where a few services are already deployed and running:

Author: I submit that we should stick to [US spelling](https://www.grammarly.com/blog/fulfil-fulfill/) so it's _fullfill_ with double _ll_

> Every time a user creates a new [service](services.md), modifies or removes an existing one, new deployment is created. Basically, the deployment is a series of steps that Marathon has to execute in order to fulfil the user request. Let's consider an example where a few services are already deployed and running:

Reviewer: ```suggestion
Every time a user creates a new [service](services.md), modifies or removes an existing one, a new deployment is created. Basically, the deployment is a series of steps that Marathon has to execute in order to fulfill the user request. Let's consider an example where a few services are already deployed and running:
```

```
/product
   /database
     /mysql
     /mongo
   /service
     /play-app
```
Now, a new service `/product/service/rails-app` needs to be deployed. To achieve this the user submits the service definition to the `/v2/apps` [Rest API](api.md) endpoint:
```
{
  "id": "/product/service/rails-app",
  "cpus": 0.3,
  "instances": 1,
  "mem": 2048,
  "cmd": ...
}
```

> If the provided definition is valid and there are no [deployment conflicts](deployments.md#forced-deployments) Marathon will return a `HTTP 201 (created)` response with a normalized service definition (including default fields that are allowed to be omitted) with an extra `deployments` field containing the deployment Id:

Reviewer: If I'm not mistaken we also set the deployment id in an extra header. Not sure if this is worth mentioning.

Author: Good point! I'll add a sentence about it.

```
{
  "id": "/product/service/rails-app",
  ...
  "deployments":[{
      "id": "3bcb15fc-737f-4a36-b185-2ae3a703239e"
    }
  ],
```
This deployment Id is a unique identifier of the process deploying the application. During the deployment Marathon will:

> - wait for Mesos to offer resources

Reviewer: Do we have articles for `offer` and `resource`?

Author: Nope

- `steps` - a series of actions to achieve the defined goal. This is being referred to as a [deployment plan](deployments.md#deployment-plan)
- `currentAction` - currently executed action of the plan

## Deployment plan
In the above example `steps` field holds the deployment plan. Our deployment plan has two deployment steps: `StartApplication` and `ScaleApplication`. Currently, following deployment steps exist:

- `StartApplication` - this step exists for backward compatibility reasons. Currently, it is doing nothing and is immediately successful
- `ScaleApplication` - is responsible for starting the defined number of tasks and waiting for them to be up and running
- `StopApplication` - stops given service by killing its tasks and waiting for the kill confirmation
- `RestartApplication` - used when modifying an existing service definition and restarting its tasks (using the new definition)

Steps are executed sequentially, however actions within one step (should there be more than one) are executed in parallel. Once all steps are executed successfully, the deployment is finished and removed from the state.

> ## Canceling a deployment

Reviewer: Nit: Do we want to capitalize headers? So `Canceling a Deployment`?

Author: I thought about it, but as far as the template goes, we don't. Should we change it? Opinions?

> Sometimes a deployment is "stuck" for some reason (see [troubleshooting](troubleshooting.md) for more information). In this case, it might be helpful to cancel the deployment, returning to the previous state. Let's cancel the above stuck deployment by sending a `DELETE /v2/deployments/3bcb15fc-737f-4a36-b185-2ae3a703239e` request. The response is `HTTP 200(ok)` but the payload might be somewhat a surprise:

Author: You can end up **without** the service after cancelation

```
{
    "version": "2018-12-05T00:28:00.584Z",
    "deploymentId": "342a6dc5-f871-4d10-a4ec-bbe21b431d58"
}
```
We see that another `deploymentId` is returned. This is because internally canceling a deployment simply creats another deployment that reverts the steps of the original one.

> **Known caveats:** Marathon has a cap on the number of concurrently running deployments. This is controlled by `--max_running_deployments` [command line argument](command-line-arguments.md). In some rare occasions, one can have the maximum number of running deployments achieved so that canceling one is not possible (since a new deployment has to be created). In this case, canceling using `?force=true` parameter aka. [forced deployment](deployments.md#forced-deployments) should be used.

Reviewer 2: ```suggestion
**Known caveats:** Marathon has a cap on the number of concurrently running deployments. This is controlled by `--max_running_deployments` [command line argument](command-line-arguments.md). In some rare cases it's possible to reach this limit so that canceling a deployments is not possible (since a new deployment has to be created). In this happens, canceling using `?force=true` parameter aka. [forced deployment](deployments.md#forced-deployments) should be used.
```

## Forced deployments

> Let's get back to our `product/service/rails-app` service deployment that is now "stuck" with slightly exaggerated resource requirements. We realize that 1k CPUs is too much and decide to update the `product/service/rails-app` service definition with reasonable CPU requirements. However, when submitting new service definition, Marathon will respond with `HTTP 209 (conflict)`:

Reviewer 2: ```suggestion
Let's get back to our `product/service/rails-app` service deployment which is now "stuck" with slightly exaggerated resource requirements. We realize that 1k CPUs is too much and decide to update the `product/service/rails-app` service definition with reasonable CPU requirements. However, when submitting the new service definition, Marathon will respond with `HTTP 209 (conflict)`:
```

```
{
    "message": "App is locked by one or more deployments. Override with the option '?force=true'",
    "deployments": [
        {
            "id": "3aa7316a-7888-427b-9af3-4179a7791c79"
        }
    ]
}
```

> This happened because already existing (stuck) deployment affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit the corrected service definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). Now, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old tasks if necessary and start new tasks.

Reviewer: `existing, in our case the stuck deployment, affects our service definition`

Author: If you'd used the `Suggestion` feature, I could've accepted it in the browser ;)

> This is due to already existing (stuck) deployment that affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit our corrected definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). This time, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old ones if necessary and start new tasks.

Reviewer 2: ```suggestion
This happened because already existing (stuck) deployment affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit the corrected service definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). Now, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old tasks if necessary and start new tasks.
```

**Known caveats:** Due to its "destructive potential" usage of forced deployments is discouraged. A forced deployment will cancel any existing conflicting deployments (there might be more than one, see [group deployments](deployments.md#group-deployments)) potentially killing more tasks than the user intended.

## Stopping a deployment

> In some rare cases a deployment might be stuck and canceling it is not an option (maybe because some tasks were already partially started and the user doesn't want to kill them). In this case, a deployment might be stopped. Stopping a deployment (as opposed to canceling a deployment) **will not return Marathon to the previous state**. Whichever tasks were already started will continue to run, the rest will not be launched. This operation will leave the service in an inconsistent state e.g. only 3/5 running tasks or 3 tasks running with the newest service definition version and 2 with the old one. **Stopping deployments should be a measure of last resort. Its usage is strongly discouraged**.

Reviewer 2: ```suggestion
In some rare cases a deployment might be stuck and canceling it is not an option (for example because some tasks were already partially started and the user doesn't want to kill them). In this case, the deployment might be stopped. Stopping a deployment (as opposed to canceling a deployment) **will not return Marathon to the previous state**. Whichever tasks were already started will continue to run, the rest will not be launched. This operation will leave the service in an inconsistent state e.g. only 3/5 running tasks or 3 tasks running with the newest service definition version and 2 with the old one. **Stopping deployments should be a measure of last resort. Its usage is strongly discouraged**.
```

# Group deployments

> Let's assume that we have already deployed databases in our cluster and now it's time to deploy the services:

Reviewer 2: ```suggestion
Let's assume that we already deployed the databases in our cluster and now it's time to deploy the services:
```

```
/database
    /mysql
    /mongo
```

> Note that we got rid of the top level `/product` group (the reason for this in the caveats section below).

Reviewer 2: An anchor link would be really nice here.

**Known caveats:** Marathon has a cap on the number of concurrently running deployments. This is controlled by `--max_running_deployments` [command line argument](command-line-arguments.md). In some rare cases it's possible to reach this limit so that canceling a deployments is not possible (since a new deployment has to be created). In this happens, canceling using `?force=true` parameter aka. [forced deployment](deployments.md#forced-deployments) should be used.

## Forced Deployments
Let's get back to our `product/service/rails-app` service deployment which is now "stuck" with slightly exaggerated resource requirements. We realize that 1k CPUs is too much and decide to update the `product/service/rails-app` service definition with reasonable CPU requirements. However, when submitting the new service definition, Marathon will respond with `HTTP 209 (conflict)`:
```
{
    "message": "App is locked by one or more deployments. Override with the option '?force=true'",
    "deployments": [
        {
            "id": "3aa7316a-7888-427b-9af3-4179a7791c79"
        }
    ]
}
```

> This happened because existing, in our case the stuck deployment, affects our service definition. In general, Marathon will only allow one deployment at a time affecting any service definition (or service definition group). Let's try to submit the corrected service definition again, however, this time using `?force=true` parameter (`POST /v2/apps?force=true`). Now, the result is `HTTP 200(ok)` and we receive a new deployment Id back. Under the hood, Marathon will cancel the previous deployment and start the new one. The new deployment will reconcile existing tasks (if any), check their service definition version, kill old tasks if necessary and start new tasks.

Reviewer 3: link to group deployments from `service group definition`?

Also should we maybe link what `reconcile` means?

> **Known caveats:** Due to its "destructive potential" usage of forced deployments is discouraged. A forced deployment will cancel any existing conflicting deployments (there might be more than one, see [group deployments](deployments.md#group-deployments)) potentially killing more tasks than the user intended.

Reviewer 3: if we recommend this , what is a recommended way to get out of the scenario described ^^^? If you have app that has e.g. resources so that it will never start? Should that just DELETE the deployment and not force override it?

> ## Stopping a Deployment

Reviewer 3: Maybe explanation of how does one stop a deployment?
